manager/controlapi/node.go

```diff
@@ -220,14 +240,19 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
 	}
 
 	if demote && s.raft != nil {
-		memberlist := s.raft.GetMemberlist()
-		for raftID, member := range memberlist {
-			if member.NodeID == request.NodeID {
-				if err := s.raft.RemoveMember(ctx, raftID); err != nil {
-					return nil, err
+		if err := s.raft.RemoveMember(ctx, member.RaftID); err != nil {
+			// Rollback to the initial Spec if we can't finalize the role
+			// change by removing the member from raft.
+			s.store.Update(func(tx store.Tx) error {
+				node = store.GetNode(tx, request.NodeID)
+				if node == nil {
+					return nil
 				}
-				break
-			}
+
+				node.Spec = initSpec
```

Comment:
> I've asked for that in a previous comment, but imho I think we should not allow the Spec change if the role change fails.

I would be okay with that in principle, but the code isn't implementing that quite correctly. Instead of not allowing it to change, it's doing a rollback on failure. But the rollback is not atomic, and if something else changes the spec in between the two transactions, their changes will be lost.

> I'm not against reverting to the either doing the rollback for the role only (Spec being partially applied), or removing this entirely.

I think what would be best for now is to not roll anything back, and leave a TODO about reattempting reconcilation of the desired state.

> Is the issue that we are querying the spec for the current role, rather than gating it through a node field?

I think this is absolutely the issue. If we had the notion of "observed role", it would solve this. Spec would be entirely under the user's control, and "observed role" would only update on a successful promotion/demotion.